utils/full_math.go

```diff
@@ -1,16 +1,56 @@
 package utils
 
 import (
+	"errors"
 	"math/big"
 
-	"github.com/daoleno/uniswapv3-sdk/constants"
+	"github.com/holiman/uint256"
 )
 
-func MulDivRoundingUp(a, b, denominator *big.Int) *big.Int {
-	product := new(big.Int).Mul(a, b)
-	result := new(big.Int).Div(product, denominator)
-	if new(big.Int).Rem(product, denominator).Cmp(big.NewInt(0)) != 0 {
-		result.Add(result, constants.One)
+var (
+	ErrMulDivOverflow = errors.New("muldiv overflow")
+	One               = big.NewInt(1)
+)
+
+// Calculates ceil(a×b÷denominator) with full precision
+func MulDivRoundingUp(a, b, denominator *uint256.Int) (*uint256.Int, error) {
+	// the product can overflow so need to use big.Int here
+	// TODO: optimize this
+	var product, rem, result big.Int
+	product.Mul(a.ToBig(), b.ToBig())
+	result.DivMod(&product, denominator.ToBig(), &rem)
+	if rem.Sign() != 0 {
+		result.Add(&result, One)
+	}
+
+	resultU, overflow := uint256.FromBig(&result)
+	if overflow {
+		return nil, ErrMulDivOverflow
+	}
+	return resultU, nil
+}
+
+// Calculates floor(a×b÷denominator) with full precision
+func MulDiv(a, b, denominator *uint256.Int) (*uint256.Int, error) {
+	// the product can overflow so need to use big.Int here
+	// TODO: optimize this follow univ3 code
+	var product, result big.Int
+	product.Mul(a.ToBig(), b.ToBig())
+	result.Div(&product, denominator.ToBig())
+
+	resultU, overflow := uint256.FromBig(&result)
+	if overflow {
+		return nil, ErrMulDivOverflow
+	}
+	return resultU, nil
+}
+
+// Returns ceil(x / y)
+func DivRoundingUp(a, denominator *uint256.Int) *uint256.Int {
```

Comment:
should we handle the 0-denominator case?

Reply:
um the original code also doesn't check for div by zero: `division by 0 has unspecified behavior, and must be checked externally` (https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/UnsafeMath.sol)

and they use raw `div`, not Solidity `/` operator, so it won't be reverted if `denominator==0`, instead it will return 0, which is the same as our code, so I think we should just keep it as is to maintain compatibility